### PR TITLE
8319896: Remove monitor deflation from final audit

### PR DESCRIPTION
**Pull request by stefank**
In [JDK-8318757](https://bugs.openjdk.org/browse/JDK-8318757) we can see that running a monitor deflation pass during a safepointed operation can interleave with the async monitor deflation, which causes various problems. The fix for that is to stop deflating monitors in the thread dump operation and instead only collect relevant monitors.

There is yet another place where we call monitor deflation from outside of the monitor deflation thread. That place is the "final audit" part, which walks over monitors and performs verification and logging just before we exit the JVM. Before the walk over the list of monitors we perform a monitor deflation pass to prune the system from "uninteresting" monitors.

I propose that we remove the monitor deflation from the final audit, and that we instead only visit "interesting" monitors (those that have an owner or "is busy"). After this change it's only the monitor deflation thread that performs monitor deflation. It is unclear to me if the final audit can actually interleave with the async monitor deflation, but this removal makes it easier to reason around monitor deflation since it is only one thread that is performing it.

This PR is dependent on #16519. I intend to run extended testing on both of these PRs over the weekend.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319896](https://bugs.openjdk.org/browse/JDK-8319896): Remove monitor deflation from final audit (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [7a82c13d](https://git.openjdk.org/jdk/pull/16605/files/7a82c13d1a67d01497b0995df573b2aab39e319e)
 * [Daniel D. Daugherty](https://openjdk.org/census#dcubed) (@dcubed-ojdk - **Reviewer**) ⚠️ Review applies to [a3df9b5d](https://git.openjdk.org/jdk/pull/16605/files/a3df9b5d196be305ffb76b32c4be2caa74eb9ee7)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16605/head:pull/16605` \
`$ git checkout pull/16605`

Update a local copy of the PR: \
`$ git checkout pull/16605` \
`$ git pull https://git.openjdk.org/jdk.git pull/16605/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16605`

View PR using the GUI difftool: \
`$ git pr show -t 16605`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16605.diff">https://git.openjdk.org/jdk/pull/16605.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16605#issuecomment-1805664380)

### CODE REVIEW

**Review by dholmes-ora — APPROVED** [review]
This seems quite reasonable to me, but the authoritative review has to come from @dcubed-ojdk :)

Thanks.

**Comment by stefank** [comment]
Thanks for reviewing. I'm going to wait for Dan's input.

**Comment by stefank** [comment]
This passes tier1-8 on Linux x64.

**Comment by stefank** [comment]
This passes tier1-3

**Comment by stefank** [comment]
> This seems quite reasonable to me, but the authoritative review has to come from @dcubed-ojdk :)

I think Dan is tied up with other things, could someone else consider reviewing this patch instead?

**Comment by dcubed-ojdk** [comment]
I finally got to a wide spot in the road where I could get back to
doing code reviews. Sorry for the delay on this review.

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up. I think I found just a couple of comment updates.

Thanks for documenting the testing that has been done.

**Comment by dcubed-ojdk** [comment]
I do like the idea of simplifying life by having only the MonitorDeflation thread be
responsible for deflating ObjectMonitors. That does simplify life and makes it much
easier to reason about monitor deflation. When I added the deflation to the
VM_ThreadDump VM operation and to the final thread auditing I was in a mode
where I was trying very hard to clean up all the garbage. While this was a useful
exercise when we were developing async monitor deflation because we were able
to find "lost" ObjectMonitors, it is no longer necessary to be so crazy about it.
The async monitor deflation code has matured for a very long time...

File: src/hotspot/share/runtime/synchronizer.cpp [hunk 1]
```diff
@@ -1614,11 +1621,8 @@ size_t ObjectSynchronizer::deflate_idle_monitors() {
   size_t deflated_count = deflate_monitor_list(current, ls, &timer);
   size_t unlinked_count = 0;
   size_t deleted_count = 0;
-  if (deflated_count > 0 || is_final_audit()) {
-    // There are ObjectMonitors that have been deflated or this is the
-    // final audit and all the remaining ObjectMonitors have been
-    // deflated, BUT the MonitorDeflationThread blocked for the final
-    // safepoint during unlinking.
+  if (deflated_count > 0) {
```
Comment (coleenp) [inline src/hotspot/share/runtime/synchronizer.cpp:1624]: There's a conditional at line 1631 is_monitor_deflation_thread, and an else that deflating idle monitors is at a safepoint.  Should there not be this conditional, ie am I reading the description right?

Comment (stefank) [reply src/hotspot/share/runtime/synchronizer.cpp:1624]: You are right. The plan in my head has been to remove the safepointed monitor deflation first and then take a pass over the code and remove the various blocks that we don't execute anymore.

**Comment by stefank** [comment]
Thanks for all the reviews! I'm going to integrate this now and will work on the follow-up cleanup.
/integrate

File: src/hotspot/share/runtime/synchronizer.cpp (hunk 1 above)
Comment (stefank) [reply src/hotspot/share/runtime/synchronizer.cpp:1624]: @coleenp I created a PR to clean this up:
https://github.com/openjdk/jdk/pull/16706

**Pull request closed without merging**